xarray/core/dataset.py

```diff
@@ -1433,11 +1449,68 @@ def __setitem__(self, key: Hashable, value) -> None:
         variable.
         """
         if utils.is_dict_like(key):
-            raise NotImplementedError(
-                "cannot yet use a dictionary as a key to set Dataset values"
-            )
+            from .dataarray import DataArray
 
-        self.update({key: value})
+            # check for consistency first
+            if isinstance(value, Dataset):
+                missing_vars = [
+                    str(name) for name in value.data_vars if name not in self.data_vars
+                ]
+                if len(missing_vars) > 0:
+                    raise KeyError(
+                        "Variables {} in new values not available in dataset!".format(
+                            ",".join(missing_vars)
+                        )
+                    )
+            for name, var in self.items():
+                missing_keys = [k for k in key.keys() if k not in var.dims]
+                if len(missing_keys) > 0:
+                    raise KeyError(
+                        "Variable {} does not contain dimensions {}!".format(
+                            name, ",".join(missing_keys)
+                        )
+                    )
+
+                # test indexing
+                try:
+                    var_k = var[key]
+                except IndexError:
+                    raise IndexError(
+                        "Indexer {} not available in variable '{}'".format(key, name)
+                    )
+
+                if isinstance(value, Dataset):
+                    val = value[name]
+                else:
+                    val = value
+                if isinstance(val, DataArray):
+                    for dim in val.dims:
+                        coord1 = var_k[dim]
+                        coord2 = val[dim]
+                        if (len(coord1) != len(coord2)) or any(
+                            coord1.values != coord2.values
+                        ):
+                            raise IndexError(
+                                "Variable '{}': dimension coordinate '{}' conflicts"
+                                " between indexed and indexing objects: "
+                                " {}\n vs.\n {}".format(name, dim, coord2, coord1)
+                            )
```

Comment:
It would be best to use `xarray.align(..., join='exact')` here rather than your own version (which may differ in some subtle ways). That's the core consistency check we use for verifying alignment in other operations.

Reply:
ok, sounds good. I used copy=False as argument. Can this anyway cause performance overhead?